module/scheduler_components/simple_remaster_manager.cpp

```diff
@@ -0,0 +1,129 @@
+#include "module/scheduler_components/simple_remaster_manager.h"
+
+#include <glog/logging.h>
+
+namespace slog {
+
+SimpleRemasterManager::SimpleRemasterManager(
+    shared_ptr<Storage<Key, Record>> storage,
+    shared_ptr<TransactionMap> all_txns)
+  : storage_(storage), all_txns_(all_txns) {}
+
+VerifyMasterResult
+SimpleRemasterManager::VerifyMaster(const TxnReplicaId txn_replica_id) {
+  auto& txn_holder = all_txns_->at(txn_replica_id);
+  auto& keys = txn_holder.KeysInPartition();
+  if (keys.empty()) {
+    return VerifyMasterResult::VALID;
+  }
+
+  auto txn = txn_holder.GetTransaction();
+  auto& txn_master_metadata = txn->internal().master_metadata();
+  if (txn_master_metadata.empty()) { // This should only be the case for testing
+    LOG(WARNING) << "Master metadata empty: txn id " << txn->internal().id();
+    return VerifyMasterResult::VALID;
+  }
+
+  // Determine which local log this txn is from. Since only single home or
+  // lock only txns, all keys will have same master
+  auto local_log_machine_id = txn_master_metadata.begin()->second.master();
+
+  // Block this txn behind other txns from same local log
+  // TODO: check the counters now? would abort earlier
+  if (blocked_queue_.count(local_log_machine_id)) {
+    blocked_queue_.at(local_log_machine_id).push_back(txn_replica_id);
+    return VerifyMasterResult::WAITING;
+  }
+
+  // Test counters
+  auto result = CheckCounters(txn_holder);
+  if (result == VerifyMasterResult::WAITING) {
+    blocked_queue_[local_log_machine_id].push_back(txn_replica_id);
+    return VerifyMasterResult::WAITING;
+  } else {
+    return result;
+  }
+}
+
+VerifyMasterResult SimpleRemasterManager::CheckCounters(TransactionHolder& txn_holder) {
+  auto& keys = txn_holder.KeysInPartition();
+  auto& txn_master_metadata = txn_holder.GetTransaction()->internal().master_metadata();
+  for (auto& key_pair : keys) {
+    auto key = key_pair.first;
+
+    // Get txn's counter
+    CHECK(txn_master_metadata.count(key))
+            << "Master metadata for key \"" << key << "\" is missing";
+    auto txn_counter = txn_master_metadata.at(key).counter();
```

Comment:
Use `[]` instead of `at`

Reply:
Seems like [] is not allowed on the const protobuf map. Could use the mutable gets, but I'd rather just remove the redundant CHECK

Reply:
Oh I didn't notice that it is a const map. `[]` automatically adds new entry so it is not allow on const object. The check is nice with the verbal explanation so if you want to keep it, it should be put at the point where we extract the keys out of the transaction in the transaction holder.